stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+                                   project=self.canonical_project)
+        dest_exists = dest.isdir()
+        if dest_exists:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise DuplicateError('Dest path already exists. Do not duplicate!')
+            folder_dest = dest
         else:
-            return self.stat()['size']
+            if dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(dest))
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        new_file_h = file_handler.clone(project=dest.canonical_project,
+                                        folder=(folder_dest.resource or '/'))
+        if not dest_exists and not dest.endswith('/'):
+            new_file_h.rename(dest.name)
+
+    def _move(self, dest):
+        """Moves the data object to a different folder within project
+
+        Args:
+            dest (path) : The destination file/folder path within same project
+
+        Raises:
+            ValueError : When attempting to move projects
+            DNAnexusError : If attempting to move across projects
+        """
+        if not self.resource:
+            raise ValueError('Cannot move project ({})'.format(self))
+        if dest.canonical_project != self.canonical_project:
+            # This can be implemented by clone and remove original
+            raise DNAnexusError('Cannot move across different projects')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        dest_exists = dest.isdir()
+        if dest_exists:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(new_dest))
+            folder_dest = dest
+        else:
+            if dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(dest))
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        file_handler.move(folder_dest.resource or '/')
+
+        if not dest_exists and not dest.endswith('/'):
+            file_handler.rename(dest.name)
+
+    def copy(self, dest, move_within_project=False):
+        """Copies data object to destination path.
+
+        Args:
+            dest (path|str): The destination file or directory.
+            move_within_project (bool): If True, make new files in DX instead of cloning.
+                Only comes into effect when both source and destination are DXPaths
+
+        Raises:
+            DNAnexusError : When cloning within same project with copy_new=False
+            DuplicateError : When cloning data object or folder to already existing path
```

Comment:
`DuplicateError` is confusing here, it's not really a duplicate, it's avoiding the creation of a duplicate.  Potentially `TargetAlreadyExistsError`?

Reply:
Yeah I guess that makes more sense. Changed to `TargetExistsError`